server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -317,14 +328,112 @@ public IndexInput openInput(String name, IOContext context) throws IOException {
+    /**
+     * Copies a list of files from the source directory to a remote based on multi-stream upload support.
+     * If vendor plugin supports uploading multiple parts in parallel, <code>BlobContainer#writeBlobByStreams</code>
+     * will be used, else, the legacy {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     * will be called.
+     *
+     * @param from    The directory for all files to be uploaded
+     * @param files   A list containing the names of all files to be uploaded
+     * @param context IOContext to be used to open IndexInput to files during remote upload
+     * @throws Exception When upload future creation fails or if {@link RemoteSegmentStoreDirectory#copyFrom(Directory, String, String, IOContext, boolean)}
+     *                   throws an exception
+     */
+    public void copyFilesFrom(Directory from, Collection<String> files, IOContext context) throws Exception {
+
+        List<CompletableFuture<UploadResponse>> resultFutures = new ArrayList<>();
+
+        for (String src : files) {
+            String remoteFilename = createRemoteFileName(src, false);
+            if (remoteDataDirectory.getBlobContainer().isMultiStreamUploadSupported()) {
+                CompletableFuture<UploadResponse> resultFuture = createUploadFuture(from, src, remoteFilename, context);
+                resultFutures.add(resultFuture);
+            } else {
+                copyFrom(from, src, src, context, false);
+            }
+        }
+
+        if (resultFutures.isEmpty() == false) {
+            CompletableFuture<Void> resultFuture = CompletableFuture.allOf(resultFutures.toArray(new CompletableFuture[0]));
+            try {
+                resultFuture.get();
+            } catch (InterruptedException | ExecutionException e) {
+                throw e;
+            }
+        }
+    }
+
+    private CompletableFuture<UploadResponse> createUploadFuture(Directory from, String src, String remoteFileName, IOContext ioContext)
+        throws Exception {
+
+        AtomicReference<Exception> exceptionRef = new AtomicReference<>();
+        long contentLength;
+        try (IndexInput indexInput = from.openInput(src, ioContext)) {
+            contentLength = indexInput.length();
+        }
+        RemoteTransferContainer remoteTransferContainer = new RemoteTransferContainer(
+            src,
+            remoteFileName,
+            contentLength,
+            true,
+            WritePriority.NORMAL,
+            (size, position) -> new OffsetRangeIndexInputStream(from.openInput(src, ioContext), size, position)
+        );
+        WriteContext writeContext = remoteTransferContainer.createWriteContext();
+        CompletableFuture<UploadResponse> uploadFuture = remoteDataDirectory.getBlobContainer().writeBlobByStreams(writeContext);
+        return uploadFuture.whenComplete((resp, throwable) -> {
```

Comment:
Should we add certain default or guesstimate timeouts so that this does not take lot of time for the upload to finish.

Reply:
another question - will whenComplete be called always (both when the upload succeeds or fails)?

Reply:
Timeouts are in place at s3 layer which is within the plugin. We can take a call on this if we need it in core right away. I am open to comments for this.

Reply:
It should be fine to upload slow as opposed to failing it  preemptively. Backpressure would protect us from indexing in case of these.